### ISSUE TITLE
Bug: Android Notification Actions not executed if pressed when app is closed

### ISSUE THREAD

**Issue by jehartzog** [Iterable/react-native-sdk#80]
### Reproduction

1. Using Android RN app, add a `customActionHandler` to this lib.
1. Using Iterable Web UI, create a push template with a push action button with `Open app` option selected.
1. Start the app, then force close it using Android swipe gesture.
1. Using Iterable Web UI, send push to device, press on the Notification Action on device

### Expected

1. App will come into focus, and execute the action that was pressed

### Actual

1. App will come into focus, action is completely ignored

### Environment:

1. `@iterable/react-native-sdk` `1.0.12` (forked by #79)
1. `react-native` `0.63.3`
1. `Android OS` `10`

### Suspected cause

1. The `iterable-android-sdk` fires a [push action](https://github.com/Iterable/iterable-android-sdk/blob/85edd3322a442b733b5bc2a19a3f1b1aa2c57591/iterableapi/src/main/java/com/iterable/iterableapi/IterablePushActionReceiver.java#L35) call on `onReceive`, but I couldn't find anywhere it was doing this check on app startup. Based on previous work we did with Android vs iOS push notifications, we needed to explicitly check for a notification once on app startup. Not sure if that is case here, just an idea.

1. Another possible idea is how the `iterable-android-sdk` work together. Since this lib receives notification action events via a 1 way listener that is registered [here](https://github.com/Iterable/react-native-sdk/blob/baea2e90c162d6bac72ab6a9f188f82cef5b5e8a/ts/Iterable.ts#L194), there may be a race condition where the event gets fired before the listener designed to catch it is actually ready. Again I didn't dive in to confirm this.

### Our Workaround

On app load we fire off [doAndroidAppLoadCheck()](https://gist.github.com/jehartzog/9490e310a07d74a546c680a27c4a339a) exactly once to make sure we've processed any notification actions that were pressed while the app is closed. It's a pretty hacky workaround, but seems to be working based on our early testing.

**Comment by vbabenkoru** [comment]
We construct a `PendingIntent` that sends a broadcast message when the push or any of the action buttons are clicked. There is a `BroadcastReceiver` in the SDK that handles those clicks.
The issue in RN SDK is that the SDK is initialized from RN `App`, and that happens much later than `Application#onCreate`. Our SDK requires that it is initialized in `Application#onCreate` to make sure all callbacks are registered by the time any `BroadcastReceiver`s are called.
In this case, the push action handler is called before the SDK is initialized, and as a result it doesn't call the url/custom action handler.

**Comment by jehartzog** [comment]
@vbabenkoru Thank you for such a quick response.

Your explanation makes sense. My follow up is that it seems `@iterable/react-native-sdk` only supports initialization from `RN App`, correct? So at this time there is not workaround to make sure that `iterable-android-sdk` is always aware that there is a `@iterable/react-native-sdk` that should be receiving these events (#81).

**Comment by vbabenkoru** [comment]
Yes. I think a 'proper' solution would be to store a pending action if the SDK is not initialized, and executing it after SDK initialization. That'd be a fix in the native SDK. We have this in the backlog, just haven't gotten around to implementing it yet.

**Comment by shzhng** [comment]
@jehartzog @vbabenkoru am i right in presuming this also afflicts push notifications where the action is **Open URL** with a deep link url? we are seeing push notifications opening the app, but although we have a custom urlHandler defined, it doesn't trigger and navigation does not happen.

**Comment by jehartzog** [comment]
@shzhng Correct. It was easier at first for us, we could return `false` from our `config.urlHandler` for Android only, and rely on [this code](https://github.com/Iterable/react-native-sdk/blob/2bdece7fbfede1250c5b1fc7ac2ca06add3ad5c5/ts/Iterable.ts#L184) to execute it. But it's complicated because you can no longer really do the full behavior you may expect to be able to do in a custom handler, as you have to work around it gets passed to your app. And this started not working as when we added in notification action handling.

We also had issues with now calling deep-links twice, so we had to add a `debounce` in there to special case that. I don't honestly have a perfect answer, except to say that these few Android issues took a long time of careful workarounds & testing to get something that was production ready.

**Comment by vbabenkoru** [comment]
@jehartzog @shzhng We've released a new version, 1.0.19, that should address this issue. It bumps the dependencies to the recent iOS/Android SDK versions where we added support for delayed initialization.

**Comment by shzhng** [comment]
@vbabenkoru we've upgraded `@iterable/react-native-sdk` to 1.0.19, and subsequently the android dependency `implementation "com.iterable:iterableapi:3.2.9"` (have you considered having this be a direct `implementation` dependency in this library rather than `api` as it currently is?)

Unfortunately the problem still persists. We are passing in a custom `urlHandler` and it still does not trigger on either the initial cold open from push notification, or from a backgrounded state where it's already loaded. We've confirmed that the same configuration does work on iOS (both cold open & backgrounded).

**Comment by shzhng** [comment]
@vbabenkoru any chance you know the issue/problem that we're running into?

**Comment by vbabenkoru** [comment]
We'll take a look at this soon, thank you for posting this.

> (have you considered having this be a direct implementation dependency in this library rather than api as it currently is?)
This is done on purpose, so that the app can access Iterable SDK APIs directly if needed. When using `implementation`, transitive dependencies are not accessible from the app code.

**Comment by vbabenkoru** [comment]
@shzhng @jehartzog Are you still experiencing this with the latest RN SDK versions?

**Comment by jehartzog** [comment]
@vbabenkoru Our workaround for Android has been stable in our production app. We have on the roadmap to upgrade to latest version of this SDK and will re-validate, but that is weeks/months away. I'll be sure to swing back if I see the issue reoccurring. Thanks again for the response and fix, we will use it soon enough.

**Comment by vbabenkoru** [comment]
@jehartzog What is the workaround?

**Comment by jehartzog** [comment]
See the bottom section of original post, I linked to a [draft version](https://gist.github.com/jehartzog/9490e310a07d74a546c680a27c4a339a) of how we hacked our way around this limitation.

**Comment by jehartzog** [comment]
@vbabenkoru We finally reached the point where we're doing testing w/ the latest lib on our new app.

Unfortunately, the issue still occurs, and we're actually seeing it happen on [iOS now too](https://github.com/Iterable/react-native-sdk/issues/130#issuecomment-849809178).

I tested 3x versions of the lib on both OS's and saw the same behavior in all cases: deep links from pushes work when the app is backgrounded or foregrounded. No links are processed when the app is terminated.

I'm afraid unless there is an easy fix, we will have to revert to downgrading to the `@iterable/react-native-sdk": "1.0.21"` we're currently using in our other production application, and bringing over the hacky workarounds we created to make it work on Android.
